### ISSUE TITLE
Infer unannotated override parameter types from the overridden method

### ISSUE THREAD

**Issue by h3har** [astral-sh/ty#4424]
## Summary

When a subclass overrides an annotated base-class method but omits annotations from the overriding method's parameters, ty infers those parameters as `Unknown`.

When the override is unambiguous, ty could use the corresponding parameter types from the specialized base-class signature as contextual inference defaults. This would preserve type information that is already available from the method being overridden.

## Reproduction

```python
from typing import reveal_type


class Base:
    def process(self, value: str, count: int) -> None:
        pass


class Child(Base):
    def process(self, value, count):
        reveal_type(value)
        reveal_type(count)
```

Run:

```console
ty check repro.py
```

With ty 0.0.75, the revealed types are:

```text
Unknown
Unknown
```

## Expected behavior

The overriding parameters should be inferred from `Base.process`:

```text
str
int
```

An explicit annotation on the overriding method should continue to take precedence. The overriding method's return type should continue to use normal return-type inference rather than being inherited from the base method.

## Suggested initial scope

A conservative initial rule could infer parameter types only when:

- the overridden base method is not overloaded;
- there is one unambiguous overridden method after normal method-resolution-order lookup;
- the child method has the same parameter names and compatible parameter kinds and shape; and
- each inherited parameter type is specialized for the subclass before it is used.

For example, specialization should apply here:

```python
from typing import Generic, TypeVar, reveal_type

T = TypeVar("T")


class Base(Generic[T]):
    def process(self, value: T) -> None:
        pass


class Child(Base[str]):
    def process(self, value):
        reveal_type(value)  # expected: str
```

Cases involving overloads, renamed parameters, incompatible parameter shapes, or ambiguous multiple inheritance could initially retain the existing `Unknown` behavior.

## Motivation

An overriding method is constrained by the callable contract established by its base class. When the child omits an annotation, using the corresponding base parameter type is a useful contextual default: it avoids duplicating annotations while retaining type checking, completion, hover information, and other language-server support within the overriding method's body.

This is particularly useful for overrides that add behavior before or after a `super()` call while otherwise preserving the base method's signature.

Whether a strict annotation-completeness rule should report that the annotations are syntactically absent is separate from which semantic types ty uses to check the method body.

Pyright documents equivalent behavior under [Parameter Type Inference](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#parameter-type-inference).

## Related issues

- [#166](https://github.com/astral-sh/ty/issues/166) tracks override compatibility checks, rather than contextual inference for omitted override annotations.
- [#1889](https://github.com/astral-sh/ty/issues/1889) is the general type-system feature overview but does not currently list this behavior.
- [#136](https://github.com/astral-sh/ty/issues/136) concerns choosing an explicitly declared local type over an inferred type, rather than inheriting parameter types from an overridden method.
- [#181](https://github.com/astral-sh/ty/issues/181) concerns contextual inference for lambda parameters, which is a related concept but a different construct.
- [Pyrefly #4742](https://github.com/facebook/pyrefly/issues/4742) is an equivalent issue filed for Pyrefly, which unlike Pyright does not support this yet (as of `1.2.0`)

## Version

```text
ty 0.0.75 (a553dcc1a 2026-08-26)
```

**Comment by sharkdp** [comment]
Thank you for reporting this.

Inferring parameters as their base-class equivalent may be too strict. Child classes can accept wider parameter types. For example, this code is perfectly fine:

```py
class Base:
    def process(self, value: str) -> None:
        ...

class Child(Base):
    def process(self, value: str | bytes) -> None:
        ...

Child().process(b"foo")
```

However, if we remove the `str | bytes` annotation, `Child`'s `value` parameter would be inferred as `str` according to this proposal, which would lead to a false positive error on the `Child().process(b"foo")` call.

So doing this would violate the [gradual guarantee](https://typing.python.org/en/latest/spec/concepts.html#the-gradual-guarantee) which ty generally tries to adhere to.

We can still leave this ticket open for discussion.

**Comment by h3har** [comment]
This is something that pyright does, which I have found to be useful in reducing boilerplate. Note that the typing spec intentionally specifies the guarantee as "[not] a strict requirement, but ... a useful guideline".

Another option that is to infer a partially static type like `str | Unknown` so that some LSP support is still available.

**Comment by sharkdp** [comment]
> Another option that is to infer a partially static type like `str | Unknown` so that some LSP support is still available.

Yes, this seems like a reasonable middle ground for me.

**Comment by carljm** [comment]
I think that inferring `str` here is too aggressive, but inferring `str | Unknown` is reasonable. Barring disagreement, let's remove `needs-decision` and call that the plan?

**Comment by jelle-openai** [comment]
I actually think inferring `str` is reasonable and probably does the right thing in the majority of cases. It might be worth trying a few variants and seeing the impact on ecosystem code.

**Comment by h3har** [comment]
> I actually think inferring `str` is reasonable and probably does the right thing in the majority of cases. It might be worth trying a few variants and seeing the impact on ecosystem code.

A third option is to make it configurable. E.g., a config that enables strict inference for unannotated override parameter types (`str`) rather than a weaker type (`str | Unknown`). But I understand that adding config for everything can become unwieldy.

**Comment by h3har** [comment]
Another option that is apparently used by pyrefly https://github.com/facebook/pyrefly/issues/4742#issuecomment-5627368413 is to only apply this for methods explicitly marked with `@override`.
